File: C++/micropuzzle36.cpp
/* Program Micropuzzle36.cpp to solve micropuzzle 36 from
   J. J. Clessa "Math and Logic Puzzles for PC Enthusiasts"
*/

#include <cmath>
#include <iostream>
#include <vector>

#include "numbertheory.h"

using std::cout;
using std::endl;

using namespace math;

const double eps = 1.0e-8;

template<typename T>
bool is_cube(T n)
{
    double x = static_cast<double>(n);
    T root = static_cast<T>(std::floor(std::cbrt(x) + eps));
    return root * root * root == n;
}

int main(void)
{
    std::vector<long long> div = divisors(2000000LL);
    for (auto d : div) {
        if (is_cube(d * d + 2000000LL)) {
            cout << "Solution: factor = " << d << endl;
        }
    }
}

File: Java/Micropuzzle36.java
/* Program Micropuzzle36.java to solve micropuzzle 36 from
   J. J. Clessa "Math and Logic Puzzles for PC Enthusiasts"
   More cubes and squares
   (uses divisors function of my NumberTheory library)
*/

public class Micropuzzle36
{
    public static boolean isCube(long n)
    {
        long root = Math.round(Math.cbrt(n));
        return root * root * root == n;
    }

    public static void main(String[] args)
    {
        final long TwoM = 2000000;
        long[] div = NumberTheory.divisors(TwoM);
        for (long d : div) {
            if (isCube(d * d + TwoM)) {
                System.out.println("Solution: factor = " + d);
            }
        }
    }
}
